ui/pages/token-details/token-details-page.js

```diff
@@ -0,0 +1,229 @@
+            <Box marginTop={1}>
+              <Identicon
+                diameter={32}
+                address={token.address}
+                image={tokenMetadata ? imagePath : token.image}
+              />
+            </Box>
+          </Box>
+          <Typography
+            margin={[4, 0, 0, 0]}
+            variant={TYPOGRAPHY.H7}
+            color={COLORS.UI4}
+            data-testid="token-details-page__token-currency-balance"
+          >
+            {tokenCurrencyBalance || ''}
+          </Typography>
+          <Typography
+            margin={[6, 0, 0, 0]}
+            variant={TYPOGRAPHY.H9}
+            color={COLORS.UI4}
+            fontWeight={FONT_WEIGHT.BOLD}
+            data-testid="token-details-page__token-contract-address-title"
+          >
+            {t('tokenContractAddress')}
+          </Typography>
+          <Box display={DISPLAY.FLEX}>
+            <Typography
+              variant={TYPOGRAPHY.H7}
+              margin={[2, 0, 0, 0]}
+              color={COLORS.BLACK}
+              overflowWrap={OVERFLOW_WRAP.BREAK_WORD}
+              className="token-details__token-address"
+              data-testid="token-details-page__token-contract-address"
+            >
+              {token.address}
+            </Typography>
+            <Tooltip
+              position="bottom"
+              title={copied ? t('copiedExclamation') : t('copyToClipboard')}
+              containerClassName="token-details__copy-icon"
+            >
+              <button
+                type="link"
+                onClick={() => {
+                  handleCopy(token.address);
+                }}
+                title=""
+                className="token-details__copyButton"
+                data-testid="token-details-page__copy-button"
+              >
+                <CopyIcon size={11} color="#037DD6" />
+              </button>
+            </Tooltip>
+          </Box>
+          <Typography
+            variant={TYPOGRAPHY.H9}
+            margin={[4, 0, 0, 0]}
+            color={COLORS.UI4}
+            fontWeight={FONT_WEIGHT.BOLD}
+            data-testid="token-details-page__token-decimal-title"
+          >
+            {t('tokenDecimalTitle')}
+          </Typography>
+          <Typography
+            variant={TYPOGRAPHY.H7}
+            margin={[1, 0, 0, 0]}
+            color={COLORS.BLACK}
+            data-testid="token-details-page__token-decimals"
+          >
+            {token.decimals}
+          </Typography>
+          <Typography
+            variant={TYPOGRAPHY.H9}
+            margin={[4, 0, 0, 0]}
+            color={COLORS.UI4}
+            fontWeight={FONT_WEIGHT.BOLD}
+            data-testid="token-details-page__current-network-title"
+          >
+            {t('network')}
+          </Typography>
+          <Typography
+            variant={TYPOGRAPHY.H7}
+            margin={[1, 0, 0, 0]}
+            color={COLORS.BLACK}
+            data-testid="token-details-page__current-network"
```

Comment:
The `data-testid` attributes are added in order to support an efficient way to do e2e testing. At the moment there is no good way to identify and 'pick-up' these elements in the e2e testing suite.

We had a conversation about it here on the Slack: https://consensys.slack.com/archives/C02GH833686/p1643203963098900 and agreed that we're going to use these attributes.

Having these attributes in the reusable components as well, will help with e2e testing in the future.

Reply:
There's the ability to find elements by text, and a number of other resources before we rely on data-test-ids.... @Gudahtt do we now prefer data-test-id over finding elements by css paths and text?

Reply:
IMO the test ID selectors are preferable to using CSS or ID-based selectors because they are less fragile and error prone. IDs introduced for use in selectors like this can break functionality (if the ID isn't unique), and using CSS selectors can be hazardous if the same class is used elsewhere without your knowledge (I saw a "false positive" e2e test once because of this problem).

Using text as the selector is nice though. I like "what the user sees"-based selectors because it makes the test easier to read. It's more fragile because the text is more likely to change, but the tradeoff is worth it for me. Calling attention to text changes can be useful too.

Reply:
I like using data-test-ids because it is less fragile, so my preferences fall on the other side of the tradeoff Mark describes.. A combination of good test descriptions, well thought out assertions and code comments can help with test readability.

Regarding the point "Calling attention to text changes can be useful too.": we could just write assertions against text when that is important.

@brad-decker Why should we not use data-test-ids?

Reply:
We should use data-testids when we cannot use a selector that selects based on visible elements on the page. Like looking for a Button with the word "Next". If we first try to use text based selectors that work the way our users are using the product then we will get better warning about changes that will impact how our users will use the app. If we use a testid for that same scenario we are essentially saying "what the button says is irrelevant, our users won't be affected by what the button says" -- but as we have learned changing the one orange button in our app to blue was enough to create a viral tweet about how MetaMask may have been hacked.

testids should be used ahead of classNames etc like @Gudahtt pointed out, but no ahead of using the text on the screen as our users will.

Reply:
>  Regarding the point "Calling attention to text changes can be useful too.": we could just write assertions against text when that is important.

we should assume all text is important, and to do so we could simply *try* to find elements by text first and if its impossible fall back to -testids. Considering that the Typography component is specifically used to display Text, this component should probably never need the testid.

Reply:
@Gudahtt @brad-decker So, to confirm, we should establish a standard in our codebase that says: all e2e tests should select elements by text rendered on screen where possible, and by test-ids where that is not possible. ?

Reply:
Oh, if we are going to talk about e2e test code standards, I'd also like to know if @PeterYinusa has any thoughts on this topic?

Reply:
@david0xd @VSaric Let's proceed with making changes according to the rule:
- If an element can be selected based on its text content, don't add data-test-ids.
- If an element that needs to be seleccted in e2e tests cannot be selected based on its text content, do add data-test-ids